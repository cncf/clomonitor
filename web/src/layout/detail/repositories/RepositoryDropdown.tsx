import classNames from 'classnames';
import { MouseEvent as ReactMouseEvent, useRef, useState } from 'react';
import { GoThreeBars } from 'react-icons/go';
import { useParams } from 'react-router-dom';

import useOutsideClick from '../../../hooks/useOutsideClick';
import ExternalLink from '../../common/ExternalLink';
import styles from './RepositoryDropdown.module.css';

interface Props {
  repoName: string;
}

const RepositoryDropdown = (props: Props) => {
  const { project, foundation } = useParams();
  const ref = useRef(null);
  const [visibleDropdown, setVisibleDropdown] = useState<boolean>(false);
  useOutsideClick([ref], visibleDropdown, () => setVisibleDropdown(false));

  return (
    <>
      <div ref={ref} className="ms-auto position-relative">
        <button
          data-testid="dropdown-btn"
          type="button"
          className={`btn btn-sm btn-primary text-white rounded-0 lh-1 ${styles.btn}`}
          onClick={(e: ReactMouseEvent<HTMLButtonElement, MouseEvent>) => {
            e.preventDefault();
            e.stopPropagation();
            setVisibleDropdown(!visibleDropdown);
          }}
        >
          <GoThreeBars />
        </button>

        <ul
          role="complementary"
          className={classNames('dropdown-menu rounded-0', styles.dropdown, { show: visibleDropdown })}
        >
          <li>
            <ExternalLink
              href={`/api/projects/${foundation}/${project}/${props.repoName}/report.md`}
              className="dropdown-item lightText"
              label="Open repository report"
              target="_self"
              onClick={() => setVisibleDropdown(false)}
            >
              <div>Get markdown</div>
            </ExternalLink>
          </li>
        </ul>
      </div>
    </>
  );
};

export default RepositoryDropdown;
