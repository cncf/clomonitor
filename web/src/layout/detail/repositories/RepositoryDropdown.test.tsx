import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ReactRouter, { BrowserRouter as Router } from 'react-router-dom';

import RepositoryDropdown from './RepositoryDropdown';

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as object),
  useParams: jest.fn(),
}));

const defaultProps = {
  repoName: 'repo',
};

describe('RepositoryDropdown', () => {
  beforeEach(() => {
    jest.spyOn(ReactRouter, 'useParams').mockReturnValue({ project: 'proj', foundation: 'cncf' });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <Router>
        <RepositoryDropdown {...defaultProps} />
      </Router>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders properly', () => {
      render(
        <Router>
          <RepositoryDropdown {...defaultProps} />
        </Router>
      );

      expect(screen.getByTestId('dropdown-btn')).toBeInTheDocument();
    });

    it('opens dropdown', async () => {
      render(
        <Router>
          <RepositoryDropdown {...defaultProps} />
        </Router>
      );

      const btn = screen.getByTestId('dropdown-btn');
      await userEvent.click(btn);

      expect(await screen.findByRole('complementary')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Open repository report' })).toBeInTheDocument();
    });
  });
});
