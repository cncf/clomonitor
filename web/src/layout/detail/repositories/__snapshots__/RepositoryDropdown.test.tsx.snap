// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryDropdown creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="ms-auto position-relative"
  >
    <button
      class="btn btn-sm btn-primary text-white rounded-0 lh-1 btn"
      data-testid="dropdown-btn"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 12 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.41 9H.59C0 9 0 8.59 0 8c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zm0-4H.59C0 5 0 4.59 0 4c0-.59 0-1 .59-1H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1h.01zM.59 11H11.4c.59 0 .59.41.59 1 0 .59 0 1-.59 1H.59C0 13 0 12.59 0 12c0-.59 0-1 .59-1z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <ul
      class="dropdown-menu rounded-0 dropdown"
      role="complementary"
    >
      <li>
        <a
          aria-label="Open repository report"
          class="link dropdown-item lightText"
          href="/api/projects/cncf/proj/repo/report.md"
          rel="noopener noreferrer"
          tabindex="-1"
          target="_self"
        >
          <div
            class="d-flex flex-row align-items-baseline"
          >
            <div>
              Get markdown
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
